### PR TITLE
Introduce libnymea-modbus

### PR DESCRIPTION
**Pull request by t-mon**
Improve tool and prepare autogeneration of connection classes

nymea-plugins-modbus pull request checklist:

- [x] Make sure the pull request's title is of format "Plugin name: Add support for xyz" or "New plugin: Plugin name"

- [x] Did you test the changes on hardware, if not (e.g. absence of required hardware), please mention a person to confirm it has been tested.

- [x] Did you update the plugin's README.md accordingly?

- [x] Did you update translations (`cd builddir && make lupdate`)?

- [x] If you added a new plugin, should it be added to nymea-plugins-all or nymea-plugins-maker?

### CODE REVIEW

File: libnymea-modbus/tools/README.md [hunk 1]
```diff
@@ -179,12 +225,12 @@ Example block:
 Change into your plugin sub directory.
 Assuming you wrote the registers.json file you can run now following command to generate your modbus class:
 
-`$ python3 ../modbus/tools/generate-connection.py -j registers.json -o . -c MyModbusConnection`
+`$ python3 ../modbus/tools/generate-connection.py -j registers.json -o . -c MyModbus`
 
 You the result will be a header and a source file called:
```
Comment (mzanetti) [inline libnymea-modbus/tools/README.md:230]: -You

File: libnymea-modbus/tools/connectiontool/modbusrtu.py [hunk 2]
```diff
@@ -63,7 +65,7 @@ def writePropertyGetSetMethodImplementationsRtu(fileDescriptor, className, regis
             elif registerDefinition['registerType'] == 'coils':
                 writeLine(fileDescriptor, '    return m_modbusRtuMaster->writeCoils(m_slaveId, %s, values);' % (registerDefinition['address']))
             else:
-                print('Error: invalid register type for writing.')
+                logger.warning('Error: invalid register type for writing.')
```
Comment (mzanetti) [inline libnymea-modbus/tools/connectiontool/modbusrtu.py:68]: Isn't there something like `logger = logging.getLogger('modbus-tools')` missing? I'm not sure, just asking as I noticed it being there in other files, but not all of them.

File: libnymea-modbus/tools/generate-connection.py [hunk 3]
```diff
@@ -482,48 +482,127 @@ def writeRtuSourceFile():
...
 if protocol == 'TCP':
+    writeTcp = True
+    writeRtu = False
+elif protocol == 'RTU':
+    writeTcp = False
+    writeRtu = True
+else:
+    # Any other value generates both classes
+    writeTcp = True
+    writeRtu = True
```
Comment (mzanetti) [inline libnymea-modbus/tools/generate-connection.py:548]: would probably have used
```
writeTcp = protocol in ["TCP", "BOTH"]
writeRtu = protocol in ["RTU", "BOTH"]
```
and not generate both files if the user makes a typo in the json.

File: nymea-plugins-modbus.pro [hunk 4]
```diff
@@ -2,7 +2,7 @@ TEMPLATE = subdirs
 
 # Note keep it ordered so the lib will be built first
 CONFIG += ordered
```
Comment (mzanetti) [inline nymea-plugins-modbus.pro:4]: every time you use `CONFIG += ordered` a kitten dies :-P

https://blog.rburchell.com/2013/10/every-time-you-configordered-kitten-dies.html

better using
`libnymea-sunspec.depends = libnymea-modbus` 
instead, that is, if the sunspec depends on the modbus lib.

The reason is that config+=ordered disables parallell building

**Review by mzanetti — CHANGES_REQUESTED** [review]
mostly ok, some smaller comments/questions

File: nymea-plugins-modbus.pro (hunk 4 above)
Comment (t-mon) [reply nymea-plugins-modbus.pro:4]: woops, must be an old copy past error...

File: libnymea-modbus/tools/connectiontool/modbusrtu.py (hunk 2 above)
Comment (t-mon) [reply libnymea-modbus/tools/connectiontool/modbusrtu.py:68]: it's in the common file globaly defined, there might be better ways to use in the entire application, I just made it work somehow for me.

File: libnymea-modbus/tools/README.md (hunk 1 above)
Comment (t-mon) [reply libnymea-modbus/tools/README.md]: Updated section: 70dc99a33f1ae34dd27815bf37d87a21507af28b

File: libnymea-modbus/tools/generate-connection.py (hunk 3 above)
Comment (t-mon) [reply libnymea-modbus/tools/generate-connection.py]: Updated logic: e1d3593dbceca21ede514d018729a41d0e58c6e0

File: nymea-plugins-modbus.pro (hunk 4 above)
Comment (t-mon) [reply nymea-plugins-modbus.pro:4]: Will be fixed in the #66 for better testing purpose.

Comment (t-mon) [reply nymea-plugins-modbus.pro:4]: https://github.com/nymea/nymea-plugins-modbus/pull/66/commits/e4a42817102e761c6d39b52faf761a853fcdd58a

**Review by mzanetti — APPROVED** [review]

**Comment by mzanetti** [comment]
merged

**Pull request closed without merging**